core/state_processor.go

```diff
@@ -739,6 +715,122 @@ func (p *StateProcessor) Process(block *types.WorkObject, batch ethdb.Batch) (ty
+
+	currentBlockHeight := header.Number(hc.NodeCtx())
+
+	// Loop through the predefined block depths
+	for _, blockDepth := range blockDepths {
+		// Ensure we can look back far enough
+		if currentBlockHeight.Uint64() <= blockDepth {
+			// Skip this depth if the current block height is less than or equal to the block depth
+			break
+		}
+
+		// Calculate the target block height by subtracting the blockDepth from the current height
+		targetBlockHeight := new(big.Int).Sub(currentBlockHeight, new(big.Int).SetUint64(blockDepth))
+
+		// Fetch the header and block at the calculated target height
+		targetHeader := hc.GetHeaderByNumber(targetBlockHeight.Uint64())
+		if targetHeader == nil {
+			return fmt.Errorf("header at height %d not found", targetBlockHeight.Uint64())
+		}
+
+		// Fetch the block at the calculated target height
+		targetBlock := hc.bc.GetBlock(targetHeader.Hash(), targetBlockHeight.Uint64())
+		if targetBlock == nil {
+			return fmt.Errorf("block at height %d not found", targetBlockHeight.Uint64())
+		}
+
+		for _, etx := range targetBlock.Body().ExternalTransactions() {
+			// Check if the transaction is a conversion transaction
+			if types.IsCoinBaseTx(etx) && etx.ETXSender().IsInQuaiLedgerScope() {
+				// Redeem all unlocked Quai for the coinbase address
+				internal, err := etx.To().InternalAddress()
+				if err != nil {
+					return fmt.Errorf("error converting address to internal address: %v", err)
+				}
+
+				lockupByte := etx.Data()[0]
+				lockup := params.LockupByteToBlockDepth[lockupByte]
+				if lockup == blockDepth {
+					balance := etx.Value()
+
+					if !statedb.Exist(internal) {
+						newAccountCreationGas := params.CallNewAccountGas(evm.Context.QuaiStateSize)
+
+						// Convert newAccountCreationGas (uint64) to *big.Int
+						newAccountCreationGasBigInt := new(big.Int).SetUint64(newAccountCreationGas)
+
+						// Check if balance is greater than or equal to newAccountCreationGasBigInt
+						if balance.Cmp(newAccountCreationGasBigInt) >= 0 {
+							// If balance >= newAccountCreationGasBigInt, proceed with subtraction
+							balance.Sub(balance, newAccountCreationGasBigInt)
+						} else {
+							// Continue processing, user has not mined enough to pay for state fee
+							continue
+						}
+					}
+
+					statedb.AddBalance(internal, balance)
+				}
+			}
+
+			if types.IsConversionTx(etx) && blockDepth == params.ConversionLockPeriod {
+				// If sender is in Qi, check for unlock in Quai
+				if etx.ETXSender().IsInQiLedgerScope() {
+					internal, err := etx.To().InternalAddress()
+					if err != nil {
+						fmt.Errorf("Error converting address to internal address: %v", err)
+					}
+
+					currentHeader := hc.CurrentHeader()
+					primeTerminusHash := currentHeader.PrimeTerminusHash()
+					primeTerminusHeader := hc.GetHeaderByHash(primeTerminusHash)
+					balance := misc.QiToQuai(primeTerminusHeader.WorkObjectHeader(), etx.Value())
```

Comment:
The conversion happens at the Qi/Quai rate at inclusion, not at the rate at the unlock time

Reply:
If it was already converted, why would I need process the conversion per your other comment?